### DISCUSSION TITLE
write register

### DISCUSSION THREAD

**Discussion by Elmatbaisaad** [aldas/modbus-tcp-client#114] [Q&A]
Hello everyone ! 
i m working on a project where i should read register's values , now i need to write on register but i didn't really understand the fc6 here:
$startAddress = 12288;
$value = 55;
$unitID = 0;
$packet = new WriteSingleRegisterRequest($startAddress, $value, $unitID); // NB: This is Modbus TCP packet not Modbus RTU over TCP!
echo 'Packet to be sent (in hex): ' . $packet->toHex() . PHP_EOL;
is the startAdresse is the adress of the register that we want to change his value ?
Thank you.

**Comment by aldas** [comment]
`$startAddress` is PDU address. These are 0 based - so first register is `$startAddress = 0;`
and `$value` must be maximum 2 bytes, signed int -32,768 to 32,767 or unsigned int 0 to 65,535

For more details about modbus addressing please read spec https://modbus.org/docs/Modbus_Application_Protocol_V1_1b3.pdf page 7  "4.4 MODBUS Addressing model"

**Reply by Elmatbaisaad** [reply]
thank you @aldas  for your answer , please how can i write on a specific adresse of a register and change value ?
Thank you!

**Reply by aldas** [reply]
You would `WriteSingleRegisterRequest` and `startAddress` is you specific register address. Depending on your addressing scheme you may need to subtract 1 from your register address number to make it into PDU address (PDU addresses are 0-based).  So for example your documentation says that something is at register 5. This means that PDU address would be 4 (5 - 1).

**Reply by Elmatbaisaad** [reply]
Ahh Thank you i'll try that !
